Changes in module 'nf-core/sourmash/sketch'
--- modules/nf-core/sourmash/sketch/main.nf
+++ modules/nf-core/sourmash/sketch/main.nf
@@ -18,8 +18,7 @@
     task.ext.when == null || task.ext.when
 
     script:
-    // required defaults for the tool to run, but can be overridden
-    def args = task.ext.args ?: "dna --param-string 'scaled=1000,k=21,k=31,k=51,abund'"
+    def args = task.ext.args ?: "dna --param-string 'scaled=1000,k=21,abund'"
     def prefix = task.ext.prefix ?: "${meta.id}"
     """
     sourmash sketch \\

************************************************************
