Changes in module 'ebi-metagenomics/bwamem2decontnobams'
--- modules/ebi-metagenomics/bwamem2decontnobams/main.nf
+++ modules/ebi-metagenomics/bwamem2decontnobams/main.nf
@@ -2,12 +2,8 @@
     tag "$meta.id"
     label 'process_high'
 
-    conda "${moduleDir}/environment.yml"
-    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
-        'https://depot.galaxyproject.org/singularity/mulled-v2-e5d375990341c5aef3c9aff74f96f66f65375ef6:2d15960ccea84e249a150b7f5d4db3a42fc2d6c3-0' :
-        'biocontainers/mulled-v2-e5d375990341c5aef3c9aff74f96f66f65375ef6:2d15960ccea84e249a150b7f5d4db3a42fc2d6c3-0' }"
+    container 'quay.io/microbiome-informatics/bwamem2:2.3'
 
-
     input:
     tuple val(meta), path(reads)
     tuple val(meta2), path(index)

************************************************************
